app/domain/authentication/authn_k8s/authentication_request.rb

```diff
@@ -0,0 +1,32 @@
+module Authentication
+  module AuthnK8s
+
+    # This class validates the k8s authentication request against the resource
+    # restrictions configured in the policy for the requesting host.
+    # It uses the 'validate_k8s_resource' command-class to validate a k8s
+    # resource against the k8s API.
+    # This command-class receives the resource's type and name as input.
+    class AuthenticationRequest
+      def initialize(namespace:, validate_k8s_resource:)
+        @namespace = namespace
+        @validate_k8s_resource = validate_k8s_resource
+      end
+
+      def valid?(restriction_name, restriction_value)
+        case restriction_name
+        when Restrictions::NAMESPACE
+          if restriction_value != @namespace
+            raise Errors::Authentication::AuthnK8s::NamespaceMismatch @namespace, restriction_value
+          end
+        when Restrictions::AUTHENTICATION_CONTAINER_NAME
```

Comment:
why is this here? don't you control the call to this method? can't we not call it with the container name annotation?

Also - this annotation is not really a restriction iirc. It just tells conjur to which container it should inject the cert. 
am i right?

Reply:
We don't control it, no.
Since this is part of the general ValidateResourceRestriction flow, and the `authentication-container-name` restriction is a valid restriction, it needs to be handled.
Basically, we need to handle any valid restriction that starts with `authn-k8s/`.
I thought of moving the container name validation here, but it is quite complex and not necessary.

Reply:
My comment is (sort of) related to [this comment](https://github.com/cyberark/conjur/pull/1934/files#r522421529):

It seems that implicit concept here is that of different _types_ of restrictions, at the very last you have "container name validation" and "everything else" -- the "everything else" being the ones we validate above.  I'm not commenting on this distinction itself... merely pointing out that the current code is, _de facto_, making that distinction.

So I think you have two choices:

1. Decide that this concept of different types of restrictions is, in fact, needed, and make it explicit.
2. Decide it's not needed and centralize the restriction logic.

Based only your comment:

> I thought of moving the container name validation here, but it is quite complex and not necessary.

I am guessing that 1. is the better road.  Assuming that is the case, the simple thing to do to avoid the empty case condition would be to first _filter_ your restrictions so that they don't contain the CONTAINER_NAME restriction, and only then call this method.

The method itself, no longer being fully general, would change its name to:

`valid_XXX_restriction?`

where `XXX` describes the valid type of restriction, ie, whatever you want to call the "everything except a container name description".

Reply:
I followed 1st suggestion.
Regarding `Valid_XXX_restriction`, I haven't done it for two reasons:
1. The container name is not a regular restriction, as @orenbm said.
2. The signature is common across all authenticators, and it seems not right to me to change it for all just to fix this single one.